### PR TITLE
Announce the new envvar policy.

### LINKED ISSUE #134: Blog post about the env variable policy
Write a blog post about the env variable policy

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Fixes #134

### CODE REVIEW

File: content/blog/2022/10/envvar-policy.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
+---
+title: Enforcing compliance in your environment variables
+authors:
+- José Guilherme Vanz
+date: 2022-11-01
+---
+
+We're glad to present to Kubewarden users the new [environment-variable-policy](https://github.com/kubewarden/environment-variable-policy) policy.
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:8]:
```suggestion
We're glad to present to Kubewarden users the new [environment-variable-policy](https://github.com/kubewarden/environment-variable-policy).
```
I think it removes repetion.

Signed-off-by: Víctor Cuadrado Juan <vcuadradojuan@suse.de>

File: content/blog/2022/10/envvar-policy.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+It's able to inspect containers, init containers, and ephemeral containers, and restricts
+their usage of environment variables by checking their names and values defined under the
+containers' [`env[*]`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core).
+As always, the policy can be found in [Github](https://github.com/kubewarden/environment-variable-policy)
+and all the  artifacts, including the BOM files, are signed with Sigstore
+
+### What is so useful about the new environment-variable policy?
+
+The recent added policy https://github.com/kubewarden/env-variable-secrets-scanner-policy allow users
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:17]:
```suggestion
The recent added [env-variable-secrets-scanner-policy](https://github.com/kubewarden/env-variable-secrets-scanner-policy) allow users
```

To remove repetition.

Signed-off-by: Víctor Cuadrado Juan <vcuadradojuan@suse.de>

File: content/blog/2022/10/envvar-policy.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+to check if there are some leaked sensitive data in the environment variables. However, until the
+environment-variable-policy there were no way to check other environment variables aspects.
+This is fixed now. In this new policy users can validate which variables name and values their
+resources should have. For that, the policy gives 4 rule operator to control the
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:21]:
```suggestion
resources should have. For that, the policy gives 4 rule operators to control the
```


Signed-off-by: Víctor Cuadrado Juan <vcuadradojuan@suse.de>

File: content/blog/2022/10/envvar-policy.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        - name: "envvar3"
+    ```
+- `allAreUsed`: The container cannot use all listed volumes at once. Checks if all of the `environmentVariables` are in the Pod/Workload resource:
+    ```yaml
+settings:
+  rules:
+    - reject: allAreUsed
+      environmentVariables:
+        - name: "envvar2"
+          value: ""
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:50]: what is the behaviour when the value is `""`?

File: content/blog/2022/10/envvar-policy.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  rules:
+    - reject: notAllAreUsed
+      environmentVariables:
+        - name: "envvar3"
+          value: "envvar3_value"
+        - name: "envvar4"
+          value: "envvar4_value"
+    ```
+
+The user is allow to define multiple rules in the policy settings. The resource must be approved by all of them in order to be accepted into the cluster.
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:65]:
```suggestion
The user is allowed to define multiple rules in the policy settings. The resource must be approved by all of them in order to be accepted into the cluster.
```

Signed-off-by: Víctor Cuadrado Juan <vcuadradojuan@suse.de>

File: content/blog/2022/10/envvar-policy.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+### An example
+
+Let's see how users can enforce that their resource have all minium set of required environment variable. For that, it possible to use the `allAreUsed` operator:
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:69]:
```suggestion
Let's see how users can enforce that their resources have a minimum set of required environment variables. For that, it's possible to use the `allAreUsed` operator:
```

Signed-off-by: Víctor Cuadrado Juan <vcuadradojuan@suse.de>

File: content/blog/2022/10/envvar-policy.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    operations:
+    - CREATE
+    - UPDATE
+	settings:
+	  rules:
+	    - reject: anyIn
+	      environmentVariables:
+	        - name: "myenvvar"
+                  value: "myvalue"
+	        - name: "myenvvar2"
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:93]: on review, it displays as tabs, instead of spaces. Maybe is worth it to use spaces, in case people copy-paste and have a different tab setting or something.

File: content/blog/2022/10/envvar-policy.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+Let's see how users can enforce that their resource have all minium set of required environment variable. For that, it possible to use the `allAreUsed` operator:
+
+```yaml
+kubectl apply -f - <<EOF
+apiVersion: policies.kubewarden.io/v1
+kind: ClusterAdmissionPolicy
+metadata:
+  name: environment-variable-policy
+spec:
+  module: ghcr.io/kubewarden/policies/environment-variable-policy:latest
```
Comment (viccuad) [inline content/blog/2022/10/envvar-policy.md:78]: I suppose we should use a version tag for the policy.

**Review by viccuad — COMMENTED** [review]
LGTM, minor nitpicks.

File: content/blog/2022/10/envvar-policy.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+containers' [`env[*]`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core).
+As always, the policy can be found in [Github](https://github.com/kubewarden/environment-variable-policy)
+and all the  artifacts, including the BOM files, are signed with Sigstore
+
+### What is so useful about the new environment-variable policy?
+
+The recent added policy https://github.com/kubewarden/env-variable-secrets-scanner-policy allow users
+to check if there are some leaked sensitive data in the environment variables. However, until the
+environment-variable-policy there were no way to check other environment variables aspects.
+This is fixed now. In this new policy users can validate which variables name and values their
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:20]: I don't **personally** like the "this is fixed now" and the way we get there. I would rephrase it be something like:

```
This policy complements the recently released <link to secret scanner blog post> policy. Both policies focus on validating the environment values provided to Kubernetes Pod objects.

In this new policy, users can validate...
```

File: content/blog/2022/10/envvar-policy.md (hunk 3 above)
Comment (flavio) [reply content/blog/2022/10/envvar-policy.md:21]: I would also use `four` instead of `4`

File: content/blog/2022/10/envvar-policy.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+- `anyIn`: Works as a deny list of your usage of volumes. It checks if any of the `environmentVariables` are in the Pod/Workload resource:
+    ```yaml
+settings:
+  rules:
+    - reject: anyIn
+      environmentVariables:
+        - name: "envvar1"
+        - name: "envvar2"
+    ```
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:32]: I was skeptical about having unordered lists entries with nested code blocks. I tried to render the document locally and it looks bad/broken:

image

I propose the following approach:

* Turn each bullet point into a header 2
* Inside of each section provide:
   * Explanation of what the rule (`anyIn`, `anyNotIn`, ... ) does. This is currently written after the `:`
   * Code block as
   * An example of that specific setting

Comment (raulcabello) [reply content/blog/2022/10/envvar-policy.md:32]: It is broken because there is a blank space before ` ```yaml`. If this space is removed the rendering is correct (see image attached) However, I would also be ok with converting each bullet point into a header 2 as flavio suggested.
Screenshot from 2022-10-24 11-55-09

File: content/blog/2022/10/envvar-policy.md (hunk 7 above)
Comment (raulcabello) [reply content/blog/2022/10/envvar-policy.md:93]: copy and paste doesn't work. This is the error I see:
```
error: error parsing STDIN: error converting YAML to JSON: yaml: line 15: found a tab character that violates indentation
```

File: content/blog/2022/10/envvar-policy.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+      - name: nginx
+        image: nginx:1.14.2
+        ports:
+        - containerPort: 80
+        env:
+          - name: myenvvar
+            value: "myvalue"
+EOF
+```
+
```
Comment (raulcabello) [inline content/blog/2022/10/envvar-policy.md:132]: It would be nice to see the rejection error message here

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers, even if the text was not ready for review,  thanks for the reviews. I've updated the example and update the text with all your comments. I've marked everything as resolved. Feel free to reopened the threads if necessary.

File: content/blog/2022/10/envvar-policy.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
+---
+title: Enforcing compliance in your environment variables
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:2]:
```suggestion
title: Enforcing compliance of container's environment variables
```

File: content/blog/2022/10/envvar-policy.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+authors:
+- José Guilherme Vanz
+date: 2022-11-01
+---
+
+We're glad to present to Kubewarden users the new [environment-variable-policy](https://github.com/kubewarden/environment-variable-policy). It's able to inspect containers, init containers, and ephemeral containers, and restricts their usage of environment variables by checking their names and values defined under the containers' [`env[*]`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core).
+
+As always, the policy can be found in [Github](https://github.com/kubewarden/environment-variable-policy) and all the  artifacts, including the BOM files, are signed with Sigstore.
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:10]: I would replace the GitHub link with a link to ArtifactHub, right to the policy page

File: content/blog/2022/10/envvar-policy.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+## What is so useful about the new environment-variable policy?
+
+This policy complements the recently released  [env-variable-secrets-scanner](https://www.kubewarden.io/blog/2022/10/env-var-secrets/) policy. Both policies focus on validating the environment values provided to Kubernetes Pod objects. In this new policy, users can validate which variables name and values their resources should have. For that, the policy gives four rule operators to control the environment variable defined in the containers. They are:
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:14]:
```suggestion
This policy complements the recently released [env-variable-secrets-scanner](https://www.kubewarden.io/blog/2022/10/env-var-secrets/) policy. Both policies focus on validating the environment values provided to Kubernetes Pod objects. In this new policy, users can validate which variables name and values their resources can have.

The policy controls the environment variable defined in the containers using these four operators:
```

File: content/blog/2022/10/envvar-policy.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+### anyIn
+
+`anyIn` works as a deny list of your usage of volumes. It checks if any of the `environmentVariables` are in the Pod/Workload resource:
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:18]: There's a copy-paste leftover: "volumes".

nitpicking: add an empty line between this paragraph and the code bloc please

File: content/blog/2022/10/envvar-policy.md (hunk 4 above)
Comment (flavio) [reply content/blog/2022/10/envvar-policy.md:52]: I don't like to explain this operator by using just one value, which is also a corner case. I would use something more "traditional", like it has been done inside of the other values.

I also think it would be worth to mention both here **and** in the policy README + metadata how `""` is being handled.

File: content/blog/2022/10/envvar-policy.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  rules:
+    - reject: allAreUsed
+      environmentVariables:
+        - name: "envvar2"
+          value: ""
+```
+
+### notAllAreUsed
+
+`notAllAreUsed` the container can use all listed volumes at once, but only one of them. Checks if all of the `environmentVariables` are not in the Pod/Workload resource
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:57]: Again, a leftover from the other article: `volumes`

File: content/blog/2022/10/envvar-policy.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    - reject: notAllAreUsed
+      environmentVariables:
+        - name: "envvar3"
+          value: "envvar3_value"
+        - name: "envvar4"
+          value: "envvar4_value"
+```
+
+
+The user is allowed to define multiple rules in the policy settings. The resource must be approved by all of them in order to be accepted into the cluster.
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:71]:
```suggestion
## Using multiple rules

The user is allowed to define multiple rules in the policy settings. The resource must be approved by all of them in order to be accepted into the cluster.
```

This is worth to be highlighted inside of a dedicated chapter, otherwise it looks like this information is specific to the `notAllAreUsed` rule.

File: content/blog/2022/10/envvar-policy.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+
+## An example
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:73]: Can we move that example right into the `allAreUsed` section and add similar ones for each rule?

I understand this would grow the article a lot, maybe we can - for the yet to be written examples - use something like that:

> For example, given the following configuration:
> 
> ```yaml
> settings:
>   rules:
>     - reject: anyIn
>       environmentVariables:
>         - name: "foo"
>           value: "bar"
>         - name: "foo2"
>           value: "bar2"
> ```
> 
> The following Pod would be rejected:
> 
> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   name: nginx
>   spec:
>     containers:
>     - name: nginx
>       image: nginx:1.14.2
>       ports:
>       - containerPort: 80
>       env:
>         - name: "foo"
>           value: "bar"
> ```

Some notes about that:

  * Save space by not showing `kubectl` invocations
  * Save space by showing only the settings of the policy, not the whole `ClusterAdmissionPolicy` definition
  * Save space by showing just the `Pod` definition, which is shorter than `Deployment` or other things

It's however nice to show that the policy can work also on `Deployment` resources, hence do not overwrite the current example.

File: content/blog/2022/10/envvar-policy.md [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        env:
+          - name: "foo"
+            value: "bar"
+EOF
+# the following error message should be updated after https://github.com/kubewarden/environment-variable-policy/pull/9
+Error from server: error when creating "deployment.yaml": admission webhook "clusterwide-environment-variable-policy.kubewarden.admission" denied the request: Resource misses at least one environment variable from the rule.
+```
+
+
+Try it out!
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:143]: Now that we have multiple sections we have to create a dedicated one for the closing words.

Maybe something like "Call to action", "Summary" or something similar...

```md
## Summary

We hope this article provided you a nice overview about the potential use cases of this policy and how to use it.
Give it a try and, as usual....
```

File: content/blog/2022/10/envvar-policy.md [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+
+As always, the policy can be found in [ArtifactHub](https://artifacthub.io/packages/kubewarden/environment-variable-policy/environment-variable-policy) and all the [artifacts](https://github.com/kubewarden/environment-variable-policy/releases), including the BOM files, are signed with Sigstore.
+
+## What is so useful about the new environment-variable policy?
+
+This policy complements the recently released [env-variable-secrets-scanner](https://www.kubewarden.io/blog/2022/10/env-var-secrets/) policy. Both policies focus on validating the environment values provided to Kubernetes Pod objects. In this new policy, users can validate which variables name and values their resources can have.
+
+The policy controls the environment variable defined in the containers using these four operators:
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:16]:
```suggestion
The policy controls the environment variable defined in the containers using the four operators described in the next sections.
```

File: content/blog/2022/10/envvar-policy.md [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+      env:
+        - name: "foo"
+          value: "bar"
+        - name: "foo2"
+          value: "bar2"
+```
+
+## Using multiple rules
+
+The user is allowed to define multiple rules in the policy settings. The resource must be approved by all of them in order to be accepted into the cluster.
```
Comment (flavio) [inline content/blog/2022/10/envvar-policy.md:249]: Can you add an example of that? This can be particularly tricky to understand

**Review by flavio — COMMENTED** [review]
Great job, I love how it looks like. I left my last requests for change

File: content/blog/2022/10/envvar-policy.md [hunk 22]
```diff
@@ -0,0 +1,325 @@
+---
+title: Enforcing compliance of container's environment variables
+authors:
+- José Guilherme Vanz
+date: 2022-10-25
+---
+
+We're glad to present to Kubewarden users the new [environment-variable-policy](https://github.com/kubewarden/environment-variable-policy). It's able to inspect containers, init containers, and ephemeral containers, and restricts their usage of environment variables by checking their names and values defined under the containers' [`env[*]`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core).
```
Comment (divya-mohan0209) [inline content/blog/2022/10/envvar-policy.md:8]:
```suggestion
We're glad to present the new [environment-variable-policy](https://github.com/kubewarden/environment-variable-policy) to Kubewarden users. With this policy, you will now be able to inspect init containers and ephemeral containers. You can also restrict their usage by reviewing the names and values defined under the containers' [`env[*]`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#envvar-v1-core) field.
```

File: content/blog/2022/10/envvar-policy.md [hunk 23]
```diff
@@ -0,0 +1,325 @@
...
+          value: "bar"
+        - name: "foo3"
+          value: "bar3"
+```
+
+
+
+### notAllAreUsed
+
+`notAllAreUsed` the container can use all listed environment variables at once, but only one of them. Checks if all of the `environmentVariables` are not in the Pod/Workload resource
```
Comment (divya-mohan0209) [inline content/blog/2022/10/envvar-policy.md:194]: This is confusing phrasing. I've suggested something. Hopefully that doesn't alter the meaning.

```suggestion
`notAllAreUsed` the container can use all listed environment variables at once, but only one at a time. Check if all of the `environmentVariables` are not in the Pod/Workload resource
```

**Review by divya-mohan0209 — CHANGES_REQUESTED** [review]
Left a few suggestions.

File: content/blog/2022/10/envvar-policy.md (hunk 23 above)
Comment (jvanz) [reply content/blog/2022/10/envvar-policy.md]: I've changes the text trying to be more clear. Please, let me know if it makes more sense to you.

File: content/blog/2022/10/envvar-policy.md (hunk 21 above)
Comment (jvanz) [reply content/blog/2022/10/envvar-policy.md:250]: Done

File: content/blog/2022/10/envvar-policy.md [hunk 24]
```diff
@@ -0,0 +1,459 @@
...
+        - containerPort: 80
+        env:
+          - name: "foo3"
+            value: "bar3"
+EOF
+deployment.apps/nginx-deployment created
+```
+
+
+## A use case example
```
Comment (jvanz) [inline content/blog/2022/10/envvar-policy.md:385]: As I've added a pretty complete example of the multiple rules in use. I decided to change this title as well. I did not remove the section completely because I think many examples help to understand what the policy is able of.

Comment (flavio) [reply content/blog/2022/10/envvar-policy.md:385]: I like it, thanks

**Review by flavio — APPROVED** [review]
LGTM. It can be deployed anytime

**Comment by flavio** [comment]
@jvanz don't forget to update the date set into the article's metadata. This has to be done right before merging this PR

**Comment by flavio** [comment]
Also, don't forget to solve the DCO warnings :smile:
